mobilecoind/src/db_crypto.rs

```diff
@@ -0,0 +1,184 @@
+
+use aes_gcm::{
+    aead::{generic_array::GenericArray, Aead, Error as AeadError, NewAead},
+    Aes256Gcm,
+};
+use blake2::{Blake2b, Digest};
+use failure::Fail;
+use std::sync::{Arc, Mutex};
+
+// Domain tag for database-wide encryption.
+pub const MOBILECOIND_DB_KEY_DOMAIN_TAG: &str = "mc_account_key";
+
+/// Possible db crypto error types.
+#[derive(Debug, Fail)]
+pub enum DbCryptoError {
+    #[fail(display = "Database is encrypted and encryption key has not yet been provided")]
+    MissingEncryptionKey,
+
+    #[fail(display = "Database encryption not enabled")]
+    EncryptionNotEnabled,
+
+    #[fail(display = "Encryption key already set")]
+    EncryptionKeyAlreadySet,
+
+    #[fail(display = "Invalid encryption key length")]
+    EncryptionInvalidKeyLength,
+
+    #[fail(display = "AEAD error: {}", _0)]
+    Aead(AeadError),
+}
+
+impl From<AeadError> for DbCryptoError {
+    fn from(src: AeadError) -> Self {
+        Self::Aead(src)
+    }
+}
+
+/// Interface for database encryption support.
+pub trait DbCryptoProvider: Clone {
+    /// Sets the password to use.
+    fn set_password(&self, password: &[u8]) -> Result<(), DbCryptoError>;
+
+    /// Clear the password.
+    fn clear_password(&self) -> Result<(), DbCryptoError>;
+
+    /// Whether a password is needed or not (allows us to tell if we're running with an encryption
+    /// backend or without).
+    fn requires_password(&self) -> bool;
+
+    /// Encrypt data.
+    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError>;
+
+    /// Decrypt data.
+    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError>;
+}
+
+/// A no-op crypto provider, used when encryption is disabled.
+#[derive(Clone, Default)]
+pub struct NoDbCryptoProvider;
+
+impl DbCryptoProvider for NoDbCryptoProvider {
+    /// Sets the password to use.
+    fn set_password(&self, _password: &[u8]) -> Result<(), DbCryptoError> {
+        Err(DbCryptoError::EncryptionNotEnabled)
+    }
+
+    /// Clear the password.
+    fn clear_password(&self) -> Result<(), DbCryptoError> {
+        Err(DbCryptoError::EncryptionNotEnabled)
+    }
+
+    /// Whether a password is needed or not (allows us to tell if we're running with an encryption
+    /// backend or without).
+    fn requires_password(&self) -> bool {
+        false
+    }
+
+    /// Encrypt data.
+    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+
+    /// Decrypt data.
+    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>, DbCryptoError> {
+        Ok(data.to_vec())
+    }
+}
+
+/// Aes crypto provider
+#[derive(Clone)]
+pub struct AesDbCryptoProvider {
+    encryption_key: Arc<Mutex<Option<Vec<u8>>>>,
+}
+
+impl Default for AesDbCryptoProvider {
+    fn default() -> Self {
+        Self {
+            encryption_key: Arc::new(Mutex::new(None)),
+        }
+    }
+}
+
+impl DbCryptoProvider for AesDbCryptoProvider {
+    /// Sets the password to use.
```

Comment:
might be a good idea to document that password must be 32 bytes, and so is expected to be a hash of a password?

Reply:
I was thinking about this before. It's a bit tricky. On one hand we could pull out the 12+32 and all that stuff. And potentially make the nonce size etc all part of the class properties. Then we could document the static variables. Otherwise we can keep the numbers close to the action which makes the maths easier in my head. I think technically we should err on the side of DRY and named constants but I am not sure what I'd actually do. Names are hard